designs/data_serialization_design.md

```diff
@@ -0,0 +1,189 @@
+# FAIR Data Serialization Design
+## Overview
+
+This document outlines the design for serializing FAIR’s core data structures, which is essential for enabling FAIR as a distributed system. It also covers how serialized data is expected to be consumed and synchronized across hosts. 
+
+This document doesn't capture details about how the serialized states from other hosts are merged/considered in the decision making.
+
+## Glossary
+
+What's a Bucket? \
+The most basic data-structure in FAIR implementatin is a bucket - this contains a probability and a last-seen time. The probability is decayed (or) improved based on some factors. For the life-time of the Structure, a client - identified by the client-ID, consistently lands on a few buckets (based on seed). The final decision is made on top of the bucket probabilities.
+
+
+## Requirements
+### Functional
+
+- **Complete**: Serialize the full data structure, including:
+    - Multi-level bucket data (probabilities, timestamps)
+    -  Configuration parameters
+    - MurmurHash seed
+    - Metadata (ID, flags)
+
+- **Portable**: Work across platforms, Go versions, and FAIR versions, with proper versioning.
+- **Compact**: Add minimal overhead for efficient transfer.
+- **Future-proof**: Support schema evolution as new parameters are added.
+
+### Non-Functional
+- **Performance**: Fast serialization/deserialization.
+- **Reliability**: Strong validation and error handling.
+- **Maintainability**: Clear versioning and schema upgrade path.
+- **Compatibility**: Backward and forward compatibility.
+
+## Serialization Format Selection
+
+Rejected options: JSON, XML, and other text formats — too verbose. \
+Custom serialization: Considered, but unnecessary given current requirements. Can revisit if protobuf overhead becomes a concern. \
+Chosen option: Protocol Buffers (protobuf), due to:
+- Binary format
+- Native support in gRPC
+- Strong typing and schema evolution
+- Built-in JSON support (for debugging)
+- Schema Design
+
+The schema has two broad parts:
+1)  Config (TrackerCfg) — defined by customers, drives tracker creation.
+2) Runtime Data — driven by traffic and execution:
+   - Consts: Initialized once, updated periodically (e.g., hash seed).
+   - State: Updated frequently on the hot path (bucket data).
+
+Accessing `State` requires the matching `Config` and `Consts`. Multiple tracker configs may exist simultaneously (multi-tenant).
+
+## Schema v1 (High-level)
+```
+package fair.data.v1;
+
+enum LevelSquashingFunction {
+  LEVEL_SQUASHING_FUNCTION_UNSPECIFIED = 0;
```

Comment:
Why do we need this catch all?

Reply:
Not sure, why i added this in the first place. I will check the code for the default behavior and make that the default.